packages/fastify-podlet-server/lib/fastify-podlet-plugin.js

```diff
@@ -50,556 +50,575 @@ const resolveAssetsBasePath = ({ base, prefix, fallback }) => {
+    const procExp = new ProcessExceptionHandlers(this.#logger);
+    procExp.closeOnExit(this.#fastify, { grace: this.#grace });
+    this.#metricStreams.push(procExp.metrics);
+  }
+
+  /**
+   * Config getter
+   */
+  get config() {
+    return this.#config;
+  }
+
+  /**
+   * Metrics object getter
+   */
+  get metrics() {
+    return this.#metrics;
+  }
+
+  /**
+   * Podlet object getter
+   */
+  get podlet() {
+    return this.#podlet;
+  }
+
+  /**
+   * We register the Podium podlet Fastify plugin here and pass it the podlet
+   * and then collect the metrics
+   */
+  podletPlugin() {
+    // @ts-ignore
+    this.#fastify.register(fastifyPodletPlugin, this.#podlet);
+  }
+
+  /**
+   * Sets up response timing metrics which hooks into Fastify via a plugin and exposes
+   * metrics via a stream which we then gather up into this.#metricStreams
+   */
+  timingMetrics() {
+    const responseTiming = new ResponseTiming({
+      timeAllRoutes: this.#timeAllRoutes,
+      groupStatusCodes: this.#groupStatusCodes,
     });
+    this.#fastify.register(responseTiming.plugin());
+    this.#metricStreams.push(responseTiming.metrics);
   }
 
   /**
-   * Collect up all metrics and expose on a .metrics property of the fastify instance
-   * which can then be piped into a consumer in server.js
+   * Sets up metric stream processing. Must be called at the end after other modules have
+   * had a chance to push metric stream objects into this.#metricStreams after which
+   * we gather up these streams and pipe them altogether and expose a single metric object
+   * to be used externally for consumption
    */
-  if (METRICS_ENABLED) {
-    const metrics = new Metrics();
-    for (const stream of metricStreams) {
+  metricStreams() {
+    /**
+     * Collect up all metrics and expose on a .metrics property of the fastify instance
+     * which can then be piped into a consumer in server.js
+     */
+    for (const stream of this.#metricStreams) {
       stream.on("error", (err) => {
-        fastify.log.error(err);
+        this.#logger.error(err);
       });
-      stream.pipe(metrics);
+      stream.pipe(this.#metrics);
     }
-    fastify.decorate("metrics", metrics);
   }
-};
+
+  /**
+   * Sets up and returns a fastify plugin with everything necessary for a running app.
+   * @returns {import("fastify").FastifyPluginAsync}
+   */
+  plugin() {
+    // return plugin
+    return async (fastify) => {
+      this.#fastify = fastify;
+
+      this.podletPlugin();
+      await this.manifestRoute();
+      this.translations();
+      await this.compression();
+      await this.processExceptionHandlers();
+      this.timingMetrics();
+
+      if (this.#config.get("app.component")) {
+        await this.contentRoute();
+        await this.fallbackRoute();
+        await this.dependenciesRoute();
+        await this.serveAssets();
+      }
+
+      this.metricStreams();
+    };
+  }
+}
 
 export default fp(async function (fastify, { config }) {
```

Comment:
this is the outer plugin which is wrapped in the "fastify-plugin" module in order to support decorators that work outside the plugin and its children